### PR TITLE
Replace RE2 dependency with built-in regex engine

### PR DESCRIPTION
**Pull request by matajoh**
Remove the RE2 third-party dependency and replace it with a header-only NFA-based regex engine (regex_engine.h) and a compatibility wrapper (TRegex in regex.h) that preserves the existing API surface.

New files:
- include/trieste/regex_engine.h: NFA-based regex engine with support for POSIX character classes, Unicode properties, capture groups, and both greedy/non-greedy quantifiers.
- include/trieste/unicode_data.h: Unicode General Category tables used by the regex engine for \p{...} property escapes.
- test/regex_engine_test.cc: ~2000 lines of unit tests covering the engine's parsing, matching, and edge cases.
- test/tregex_benchmark.cc: benchmark for the TRegex wrapper layer.
- test/regex_engine_benchmark.cc: opt-in comparative benchmark against RE2 (requires TRIESTE_BUILD_REGEX_BENCHMARK=ON).

Changes to existing files:
- CMakeLists.txt: remove RE2 FetchContent and link dependency; add TRIESTE_BUILD_REGEX_BENCHMARK option that fetches RE2 only when explicitly enabled.
- include/trieste/regex.h: replace RE2 wrapper with TRegex class providing the same Arg/Match/PartialMatch interface.
- include/trieste/parse.h, rewrite.h: switch from RE2 to TRegex.
- include/trieste/utf8.h: extend UTF-8 utilities used by the engine.
- parsers/yaml/reader.cc, writer.cc: minor adjustments for the new regex interface.

This eliminates a significant build-time dependency while keeping the library fully header-only. Here is a benchmark comparison for various Trieste regular expressions comparing TRegex with RE2:

regex_benchmark_vs_re2

### CODE REVIEW

File: include/trieste/regex.h [hunk 1]
```diff
@@ -2,60 +2,584 @@
...
+      init();
+    }
+
+    explicit TRegex(std::string_view pattern) : pattern_(pattern)
+    {
+      init();
+    }
+
+    explicit TRegex(const char* pattern)
+    : pattern_(pattern == nullptr ? "" : pattern)
```
Comment (achamayou) [inline include/trieste/regex.h]: Perhaps this is necessary for RE2, in which case ignore what follows. Throwing an `std::invalid_argument` may be tempting here, rather than converting to an empty pattern?

File: include/trieste/unicode_data.h [hunk 2]
```diff
@@ -0,0 +1,5212 @@
...
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <string_view>
+#include <utility>
+
+namespace trieste::unicode
+{
+  using range_t = std::pair<uint32_t, uint32_t>;
```
[the remark this answers is not in the document]
Comment (matajoh) [reply include/trieste/unicode_data.h]: I've added a shared header that defines `rune_t`, which is a small change but resolves this nicely.
